src/pyadjoint/adjoint_source.py

```diff
@@ -140,6 +148,70 @@ def _write_specfem(self, buf, time_offset):
 
         np.savetxt(buf, to_write)
 
+    def write_to_asdf(self, ds, time_offset, **coordinates):
+        """
+        Writes the adjoint source to an ASDF file.
+        Note: For now it is assumed SPECFEM will be using the adjoint source
+
+        :param ds: The ASDF data structure read in using pyasdf.
+        :type ds: str
+        :param time_offset: The temporal offset of the first sample in seconds.
+        :type time_offset: float
+        :param latitude: If given, the latitude of the adjoint source.
+        :type latitude: float
+        :param longitude: If given, the longitude of the adjoint source.
+        :type longitude: float
+        :param elevation: If given, the elevation of the adjoint source
+            in meters.
+        :type elevation: float
+
+        .. rubric:: SPECFEM
+
+        SPECFEM requires one additional parameter: the temporal offset of the
+        first sample in seconds. The following example sets the time of the
+        first sample in the adjoint source to ``-10``.
+
+        >>> adj_src.write_to_asdf(ds, time_offset=-10,
+        ...               latitude=-17.2, longitude=20.5, elevation=2.0)
+        """
+        # Convert the adjoint source to SPECFEM format
+        l = len(self.adjoint_source)
+        specfem_adj_source = np.empty((l, 2))
+        specfem_adj_source[:, 0] = np.linspace(0, (l - 1) * self.dt, l)
+        specfem_adj_source[:, 1] = time_offset
+        specfem_adj_source[:, 1] = self.adjoint_source[::-1]
+
+        tag = "%s_%s_%s" % (self.network, self.station, self.component)
+        min_period = self.min_period
+        max_period = self.max_period
+        component = self.component
+        station_id = "%s.%s" % (self.network, self.station)
+
+        try:
+            latitude = coordinates['latitude']
+            longitude = coordinates['longitude']
+            elevation_in_m = coordinates['elevation']
+        except:
+            latitude = \
+                ds.waveforms.self.network_self.station.coordinates['latitude']
+            longitude = \
+                ds.waveforms.self.network_self.station.coordinates['longitude']
+            elevation_in_m = ds.waveforms.\
+                self.network_self.station.coordinates['elevation_in_m']
```

Comment:
While correct, the trick with `**coordinates` is pretty unusual. Additional keyword arguments are usually named `**kwargs` and passed to any underlying functions. This might be useful here to pass additional arguments to the function actually writing the data to the data set (there are none right now but who knows - there might be some at some point.)

I think its slightly cleaner to do it like this (I also don't think the previous way to extract coordinates from an ASDF did work):

```python
def write_to_asdf(self, ds, time_offset, coordinates=None):

    # Import here to not have a global dependency on pyasdf.
    from pyasdf.exceptions import NoStationXMLForStation

    ...

    if coordinates:
        # If given, all three coordinates must be present.
        if {"latitude", "longitude", "elevation_in_m"}.difference(
                set(coordinates.keys())):
            raise ValueError(
                "'latitude', 'longitude', and 'elevation_in_m' must be given")
    else:
        try:
            coordinates = ds.waveforms[
                "%s.%s" % (self.network, self.station)].coordinates
        except NoStationXMLForStation:
            raise ValueError("Coordinates must either be given directly or "
                             "already be part of the ASDF File")

    # Safeguard against funny types in the coordinates dictionary.
    latitude = float(coordinates["latitude"])
    longitude = float(coordinates["longitude"])
    elevation_in_m = float(coordinates["elevation_in_m"])

    ...
```